Find scopes for anonymous variables

smowton opened the pull request:
This adds scope information for Java anonymous locals based on their actual usage (e.g. a notionally function-scoped anonymous local variable may be bounded by a smaller lexical scope if it is only used in a particular subgraph of the CFG), achieving similar benefits to interpreting the local variable table when one isn't present (because the class file was compiled without -g, or was subsequently stripped). We may or may not want to do this in general, but we found it useful for taint analysis which likes variables to be declared dead as soon as is practical.

smowton commented:
@kroening ? This can't be moved right now anyhow, needs its predecessors to land first

mgudemann commented on src/java_bytecode/java_bytecode_convert_method_class.h:
```diff
@@ -0,0 +1,180 @@
+  symbol_exprt check_stub_function(const irep_idt& symname,
+                                   const irep_idt& basename,
+                                   const irep_idt& prettyname,
+                                   const typet& fntype);
```
please run cpplint on this file

smowton commented:
@mgu that particular line isn't in this PR (see real diff link above), but have fixed some other cpplint stuff. There's a couple of false alarms (mistaking a template argument comma for a real one, for example).

tautschnig commented:
Would it be possible to get this rebased to get rid of those commits that have already been merged? That might help further reviewing. Thanks!

smowton commented:
The immediate predecessor PR just got created (https://github.com/diffblue/cbmc/pull/458)-- will rebase this after that's done.

smowton commented:
Rebased and added a test. @kroening should probably opine on whether we want this in master. and/or if it needs to be controlled by a command-line parameter though.

tautschnig commented:
@smowton Why would one not want this by default? Is there anything that could actually go wrong (except for any remaining bugs)?

smowton commented:
Not that it could go wrong, but is does increase the size of the goto program due to injecting extra `dead` statements. Depending on the client analysis, they might prefer if locals like this all die as part of the stack frame pop rather than being eagerly weeded out as they leave scope.

tautschnig commented:
Since this change is nicely encapsulated in one commit, I'd wait for such a client analysis to step up, at which point (if worse comes to worse) 7eb318e could be reverted.

tautschnig commented:
I believe this has broken cbmc-java/inferlexicalscope1 - a proper analysis will follow.

smowton commented:
Indeed, but I've just checked and it is not broken in the "donor" branch, only in master. Working on this now...

smowton commented:
The core problem is that in master this leads to duplicate variable declarations. I suspect a dubious merge.

tautschnig commented:
Ok, I can submit a simple fix I believe - just adding whitespace in test.desc.

smowton commented:
I wouldn't, the output is genuinely incorrect. I should have something within the hour...

smowton commented:
It was indeed a bad merge from #425. Fix submitted as https://github.com/diffblue/cbmc/pull/483
